[cutlass backend] Do not change dtype of GEMM template

henrylhtsang opened the pull request:
I think this is a change in the right direction.

Right now, when we try to find a cutlass gemm, we generate bunch of gemm templates, and filter out those that don't fix. For example, if we are doing bf16 x bf16 matmul, the gemm template for fp32 x fp32 is generated and filtered out.

However, for the dtype of bias, we would attempt to modify the dtype of the gemm template. I think this is a bad idea, since (1) the usable template is also being generated, and (2) this messes with the configuration name of the template.

I tested this offline. There isn't much difference in performance. However, with instantiation level 2222, I noticed way less "C++ compile error". This is probably due to using the right template?

Follow-ups are needed:
1. benchmark and dashboard
2. check our logic for setting alignment 

with my change
https://www.internalfb.com/intern/paste/P1729604119/

without my change
https://www.internalfb.com/intern/paste/P1729624806/


Differential Revision: D69085556




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

ColinPeppler commented on torch/_inductor/codegen/cuda/gemm_template.py:
```diff
@@ -805,7 +829,8 @@ def filter_op(
 
         # Set bias layout and alignment.
-        if not self._set_bias_layout_and_alignment(op):
+        status = self._set_bias_layout_and_alignment(op)
+        if not status:
```
imo `_can_set_bias_layout_and_alignment(op)` souunds more clear.

henrylhtsang replied:
> imo `_can_set_bias_layout_and_alignment(op)` souunds more clear.

it is some inheritance, the method exists in the base class already...

ColinPeppler replied:
is it easily changable? the naming always confused me when I first read the code, but not a big deal

ColinPeppler commented on torch/_inductor/codegen/cuda/gemm_template.py:
```diff
@@ -749,21 +755,37 @@ def filter_op(
+            ):
+                return None
+        if not cutlass_utils.dtype_match(
+            self.output_node.get_layout().dtype, op.D.element
```
this could go with the A, B, accum dtype check for conciseness

henrylhtsang replied:
I was hoping to do them in order of A, B, acc, C, D. What do you think

henrylhtsang commented:
@pytorchbot merge -i

henrylhtsang commented:
@pytorchbot revert -m "mistake on logging "

henrylhtsang commented:
@pytorchbot revert -m "mistake on logging" -c ignoredsignal

henrylhtsang committed:
[cutlass backend] Do not change dtype of GEMM template (#146877)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/146877

Test Plan: https://www.internalfb.com/intern/testinfra/testrun/281475379071857

Differential Revision: D69085556

henrylhtsang commented:
@pytorchbot merge -i

henrylhtsang commented:
@pytorchbot revert -m "let me resubmit "

henrylhtsang commented:
@pytorchbot revert -m "let me resubmit " -c ignoredsignal
